language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs

```diff
@@ -54,89 +56,100 @@ nonZeroArg f x y = EvaluationSuccess $ f x y
 -- | Apply a function with a known 'TypeScheme' to a list of 'Constant's (unwrapped from 'Value's).
 -- Checks that the constants are of expected types.
 applyTypeSchemed
-    :: forall e m a r. (MonadError (ErrorWithCause e) m, AsUnliftingError e, AsConstAppError e)
-    => TypeScheme a r -> a -> [Value TyName Name ()] -> EvaluateConstApp m ()
-applyTypeSchemed = go where
+    :: forall e m a exA r. (MonadError (ErrorWithCause e) m, AsUnliftingError e, AsConstAppError e, SpendBudget m)
+    => StagedBuiltinName -> TypeScheme a exA r -> a -> exA -> [Value TyName Name ExMemory] -> EvaluateConstApp m ExMemory
+applyTypeSchemed name = go where
     go
-        :: forall b.
-           TypeScheme b r
+        :: forall b exB.
+           TypeScheme b exB r
         -> b
-        -> [Value TyName Name ()]
-        -> EvaluateConstApp m ()
-    go (TypeSchemeResult _)        y args =
+        -> exB
+        -> [Value TyName Name ExMemory]
+        -> EvaluateConstApp m ExMemory
+    go (TypeSchemeResult _)        y _ args =
+        -- TODO: The costing function is NOT run here. Might cause problems if there's never a TypeSchemeArrow.
         case args of
-            [] -> pure . ConstAppSuccess $ makeKnown y    -- Computed the result.
+            -- TODO is `withMemory` a good idea here?
+            [] -> pure . ConstAppSuccess $ withMemory $ makeKnown y    -- Computed the result.
             _  -> throwingWithCause _ConstAppError        -- Too many arguments.
-                    (ExcessArgumentsConstAppError args)
+                    (ExcessArgumentsConstAppError $ void <$> args)
                     Nothing
-    go (TypeSchemeAllType _ schK)  f args =
-        go (schK Proxy) f args
-    go (TypeSchemeArrow _ schB)    f args = case args of
+    go (TypeSchemeAllType _ schK)  f exF args =
+        go (schK Proxy) f exF args
+    go (TypeSchemeArrow _ schB)    f exF args = case args of
         []          -> pure ConstAppStuck                 -- Not enough arguments to compute.
         arg : args' -> do                                 -- Peel off one argument.
             -- Coerce the argument to a Haskell value.
-            x <- readKnownM arg
+            x <- readKnownM $ void arg
             -- Apply the function to the coerced argument and proceed recursively.
-            go schB (f x) args'
+            case schB of
+                (TypeSchemeResult _) -> do
+                    let
+                        budget :: ExBudget
+                        budget = (exF x)
+                    lift $ spendBudget (BBuiltin name) arg budget
```

Comment:
Why do we have only the last `arg` here? Can we either have all of them or none? Then you can write

```haskell
    go (TypeSchemeResult _)        y exB args =
        case args of
            -- TODO is `withMemory` a good idea here?
            [] -> do
                lift $ spendBudget (BBuiltin name) INSERT_SOMETHING_HERE exB
                pure . ConstAppSuccess $ withMemory $ makeKnown y    -- Computed the result.
            _  -> throwingWithCause _ConstAppError        -- Too many arguments.
                    (ExcessArgumentsConstAppError $ void <$> args)
                    Nothing
    <...>
    go (TypeSchemeArrow _ schB)    f exF args = case args of
        <...>
            x <- readKnownM $ void arg
            -- Apply the function to the coerced argument and proceed recursively.
            go schB (f x) (exF x) args'
```

which is certainly better.

Reply:
From what I understand, the cost may occur on `(f x)`, depending on laziness. Maybe we'll have to make this more explicit even?

Reply:
Only on the last argument? I.e. in `f x y z` only `z` is important, not `x` and `y`?

Reply:
From what I understand, once we're at `TypeSchemeResult`, the result has already been computed. So when `schB` is `TypeSchemeResult` (the next one in the chain), the result will be computed in this reduction.

Reply:
Ah, so you want to avoid computing the result, because it might be too costly? That makes sense. I think we should make use of Haskell's laziness though. And for example pass `Lazy a` for the meaning of a builtin rather than just `a` (where `data Lazy a = Lazy a`). It's most likely that we're already lazy, but I'd probably make laziness explicit still.

Reply:
Considered it, but this piece of code is a bit non-standard at one place, whereas with `Lazy` we'd have to change it anywhere, so I prefer the code as is.

Reply:
It's non-standard in one place and doesn't handle one corner case at all (the no-arrow case: a name for a constant). Which might be fine, I'm not sure.

There is another option: instead of applying the meaning of a builtin to arguments we can just collect all the arguments using something like

```haskell
data Application a where
    Pure :: a -> Application a
    Apply :: Application (a -> b) -> a -> Application b

runApplication :: Application a -> a
```

and only reduce after making sure the cost is affordable.

Reply:
Simon and I discussed something like that approach at some point, but I don't remember what we concluded about it...